File: src/1829.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *getMaximumXor(int *nums, int numsSize, int maximumBit, int *returnSize) {
    int *pRetVal = NULL;

    (*returnSize) = 0;
    pRetVal = (int *)malloc(numsSize * sizeof(int));
    if (pRetVal == NULL) {
        perror("malloc");
        return pRetVal;
    }
    memset(pRetVal, 0, (numsSize * sizeof(int)));

    int xorProduct = 0;
    for (int i = 0; i < numsSize; i++) {
        xorProduct = xorProduct ^ nums[i];
    }

    int mask = (1 << maximumBit) - 1;
    for (int i = 0; i < numsSize; i++) {
        pRetVal[(*returnSize)++] = (xorProduct ^ mask);
        // remove last element
        xorProduct = xorProduct ^ nums[numsSize - 1 - i];
    }

    return pRetVal;
}

int main(int argc, char **argv) {
#define MAX_SIZE (int)(1e5)
    struct testCaseType {
        int nums[MAX_SIZE];
        int numsSize;
        int maximumBit;
        int returnSize;
    } testCase[] = {{{0, 1, 1, 3}, 4, 2, 0}, {{2, 3, 4, 7}, 4, 3, 0}, {{0, 1, 2, 2, 5, 7}, 6, 3, 0}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: nums = [0,1,1,3], maximumBit = 2
     *  Output: [0,3,2,3]
     *
     *  Input: nums = [2,3,4,7], maximumBit = 3
     *  Output: [5,2,6,5]
     *
     *  Input: nums = [0,1,2,2,5,7], maximumBit = 3
     *  Output: [4,3,6,4,6,7]
     */

    int *pAnswer = NULL;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: nums = [");
        for (j = 0; j < testCase[i].numsSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].nums[j]);
        }
        printf("], maximumBit = %d\n", testCase[i].maximumBit);

        pAnswer =
            getMaximumXor(testCase[i].nums, testCase[i].numsSize, testCase[i].maximumBit, &testCase[i].returnSize);
        printf("Output: [");
        for (j = 0; j < testCase[i].returnSize; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", pAnswer[j]);
        }
        printf("]\n");

        printf("\n");

        free(pAnswer);
        pAnswer = NULL;
    }

    return EXIT_SUCCESS;
}

File: src/1829.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def getMaximumXor(self, nums: List[int], maximumBit: int) -> List[int]:
        retVal = []

        numsSize = len(nums)

        xorProduct = 0
        for i in range(numsSize):
            xorProduct = xorProduct ^ nums[i]

        mask = (1 << maximumBit) - 1
        for i in range(numsSize):
            retVal.append(xorProduct ^ mask)
            # remove last element
            xorProduct = xorProduct ^ nums[numsSize - 1 - i]

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums, maximumBit in zip([[0, 1, 1, 3], [2, 3, 4, 7], [0, 1, 2, 2, 5, 7]], [2, 3, 3]):
            # /* Example
            #  *  Input: nums = [0,1,1,3], maximumBit = 2
            #  *  Output: [0,3,2,3]
            #  *
            #  *  Input: nums = [2,3,4,7], maximumBit = 3
            #  *  Output: [5,2,6,5]
            #  *
            #  *  Input: nums = [0,1,2,2,5,7], maximumBit = 3
            #  *  Output: [4,3,6,4,6,7]
            #  */
            logging.info("Input: nums = %s, maximumBit = %s", nums, maximumBit)

            retVal = pSolution.getMaximumXor(nums, maximumBit)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
